File: network/src/main/kotlin/org/bitlap/network/core/operation/OperationManager.kt
package org.bitlap.network.core.operation

import cn.hutool.core.util.ServiceLoaderUtil
import org.bitlap.network.BSQLException
import org.bitlap.network.core.OperationType
import org.bitlap.network.core.Session

/**
 * Lifecycle management of operations.
 *
 * @author 梦境迷离
 * @since 2021/9/5
 * @version 1.0
 */
class OperationManager {

    private val operationFactory = ServiceLoaderUtil.loadFirst(OperationFactory::class.java)!!
    private val handleToOperation: MutableMap<OperationHandle, Operation> = mutableMapOf()

    /**
     * Create an operation for the SQL and execute it. For now, we put the results in memory by Map.
     */
    fun newExecuteStatementOperation(
        parentSession: Session,
        statement: String,
        confOverlay: Map<String, String>?
    ): Operation {
        val operation = operationFactory.create(parentSession, OperationType.EXECUTE_STATEMENT, true)
        operation.confOverlay = confOverlay
        operation.statement = statement
        operation.run()
        addOperation(operation)
        return operation
    }

    @Synchronized
    fun addOperation(operation: Operation) {
        handleToOperation[operation.opHandle] = operation
    }

    @Synchronized
    fun getOperation(operationHandle: OperationHandle): Operation {
        return handleToOperation[operationHandle]
            ?: throw BSQLException("Invalid OperationHandle: $operationHandle")
    }
}

File: network/src/main/scala/org/bitlap/net/operation/OperationManager.scala
package org.bitlap.net.operation

import cn.hutool.core.util.ServiceLoaderUtil
import org.bitlap.net.BSQLException
import org.bitlap.net.handles.OperationHandle
import org.bitlap.net.operation.operations.Operation
import org.bitlap.net.session.Session

import scala.collection.mutable

/**
 *
 * @author 梦境迷离
 * @since 2021/11/20
 * @version 1.0
 */
class OperationManager {

  private val operationFactory: OperationFactory = ServiceLoaderUtil.loadFirst(classOf[OperationFactory])
  private val handleToOperation: mutable.HashMap[OperationHandle, Operation] = mutable.HashMap[OperationHandle, Operation]()

  /**
   * Create an operation for the SQL and execute it. For now, we put the results in memory by Map.
   */
  def newExecuteStatementOperation(parentSession: Session, statement: String, confOverlay: Map[String, String] = Map.empty): Operation = {
    val operation = operationFactory.create(parentSession, OperationType.EXECUTE_STATEMENT, hasResultSet = true)
    confOverlay.foreach(kv => operation.confOverlay.put(kv._1, kv._2))
    operation.setStatement(statement)
    operation.run()
    addOperation(operation)
    operation
  }

  def addOperation(operation: Operation) {
    this.synchronized {
      handleToOperation.put(operation.opHandle, operation)
    }
  }

  def getOperation(operationHandle: OperationHandle): Operation = {
    this.synchronized {
      val op = handleToOperation.getOrElse(operationHandle, null)
      if (op == null) {
        throw BSQLException("Invalid OperationHandle: $operationHandle")
      } else {
        op
      }
    }
  }
}
